### PR TITLE
[CSE-231] 'Address does not exist' error for existing addresses.

### PR DESCRIPTION
**Pull request by ksaric**
https://issues.serokell.io/issue/CSE-231

Uncomment the necessary lines in the test, run the test, and see how it fails.

Before:

`http://localhost:8100/api/addresses/summary/DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda`

`{"Left":"Address length is not equal to 32, can't be redeeming pk"}`

After:

`http://localhost:8100/api/addresses/summary/DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda`

`{"Right":{"caAddress":"DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda","caType":"CPubKeyAddress","caTxNum":2,"caBalance":{"getCoin":"0"},"caTxList":[{"ctbId":"ac439284933ba4a2bc81e5e8995c4a0c0a39256884d905e74ffcdf7ba9b257e7","ctbTimeIssued":1506329611,"ctbInputs":[["DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda",{"getCoin":"1000000"}]],"ctbOutputs":[["DdzFFzCqrhsgHx6CUkwaLwABsZLtFZpuQsTbXpMkinerzLPCX9aNdpTVvj4P7uEmytxD8wW34y1A3f7awj1vGQMU3dBjv56619WXLjCo",{"getCoin":"329106"}],["DdzFFzCqrhsuXWViKNnvsC5NdWhT4NEw5FeDtzdm9WbzzDTTTrFSjRAPhiPhE258dt7hfHmQAHo6PNEkfzXZ7ZrF4WAiLYqPCmgUrc5G",{"getCoin":"500000"}]],"ctbInputSum":{"getCoin":"1000000"},"ctbOutputSum":{"getCoin":"829106"}},{"ctbId":"dbc87d198fc9f6067c8e65f89da60a944e1ef52f0c539d980ebbc95f3dc71b65","ctbTimeIssued":1506329391,"ctbInputs":[["Ae2tdPwUPEZ8fbeSrt7u4SfYF3GDCMBmqX4KaKygHjjcm8AjJH4R9Nbpp6u",{"getCoin":"1000000"}]],"ctbOutputs":[["DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda",{"getCoin":"1000000"}]],"ctbInputSum":{"getCoin":"1000000"},"ctbOutputSum":{"getCoin":"1000000"}}]}}`

Also you can run the `mainnet` and try it there.

### CODE REVIEW

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 1]
```diff
@@ -711,8 +715,31 @@ topsortTxsOrFail f =
 
 -- | Deserialize Cardano or RSCoin address and convert it to Cardano address.
 -- Throw exception on failure.
-cAddrToAddr :: MonadThrow m => CAddress -> m Address
-cAddrToAddr cAddr@(CAddress rawAddrText) =
+-- cAddrToAddr :: MonadThrow m => CAddress -> m Address
```
Comment (akegalj) [inline explorer/src/Pos/Explorer/Web/Server.hs]: @ksaric what's the point of having these lines left in comments ?

Comment (ksaric) [reply explorer/src/Pos/Explorer/Web/Server.hs]: @akegalj https://github.com/input-output-hk/cardano-sl/pull/1646#discussion_r140972742

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 2]
```diff
@@ -724,15 +728,14 @@ cAddrToAddr cAddr@(CAddress rawAddrText) =
...
-            unless (BS.length addr == 32) $
-                throwM badAddressLength
-            pure $ makeRedeemAddress $ redeemPkBuild addr
+            if (BS.length addr /= 32)
+                then badAddressLength
+                else makeRedeemAddress $ redeemPkBuild addr
         Nothing ->
             -- cAddr is in Cardano address format
-            either badCardanoAddress pure (fromCAddress cAddr)
+            error "Undecoded address"
```
Comment (Pastafarianist) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Please use `throwM` here as well.

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 3]
```diff
@@ -367,7 +368,10 @@ getAddressSummary
     => CAddress
     -> m CAddressSummary
 getAddressSummary cAddr = do
-    addr <- cAddrToAddr cAddr
+    -- let addr = cAddrToAddr cAddr
```
Comment (Pastafarianist) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Please don't leave commented-out code.

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 4]
```diff
@@ -367,7 +368,10 @@ getAddressSummary
...
+    addr <- case fromCAddress cAddr of
```
Comment (Pastafarianist) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Why did you get rid of `cAddrToAddr` here?

**Comment by ksaric** [comment]
@Pastafarianist, please note that this PR is identical to this one - https://github.com/input-output-hk/cardano-sl/pull/1646 and that one contains much more information.
Specifically, I think this is what you are after - https://github.com/input-output-hk/cardano-sl/pull/1646#issuecomment-332177456

**Comment by Pastafarianist** [comment]
@ksaric, basically your changes undo https://issues.serokell.io/issue/CSE-161. AFAIK that't not OK. @akegalj could you confirm that we still want to support searching for RSCoin addresses?

Apparently the problem is caused by the fact that currently Cardano addresses are serialized in such way that they can be decoded as base64.

@gromakovsky or @volhovm, could you say whether
1) Cardano addresses can be deserialized with a base64 decoder
2) it is guaranteed that they will never deserialize to a 32-byte value?

If (2) is true, then we can add a quick hack to preserve the ability to search for RSCoin addresses by checking whether the address can be decoded to a 32-byte value via base64. If it can, then it is definitely an RSCoin address; otherwise, it's a Cardano address.

If (2) is false, then we should introduce some modifications to Explorer — specifically, if the address deserializes to a 32-byte value, we should check whether the DB contains information on 1) this deserialization 2) the result of a call to `fromCAddress` on the original value. This is more difficult to implement quickly, and should probably be done post-launch.

**Comment by akegalj** [comment]
@ksaric we have to have these changes that you made in 1.0 . You should probably cherry pick these last few commits to https://github.com/input-output-hk/cardano-sl/pull/1664

**Pull request merged**